subway_prediction_ml/actuals_adder.py

```diff
@@ -0,0 +1,68 @@
+import datetime
+import os
+import pandas as pd
+import pytz
+
+from sklearn.base import BaseEstimator, TransformerMixin
+
+
+class ActualsAdder(BaseEstimator, TransformerMixin):
+    def __init__(
+        self,
+        actuals_path=os.path.join("datasets", "pa_datapoints.csv")
+    ):
+        self.actuals_path = actuals_path
+
+    def fit(self, X, y=None):
+        return self
+
+    # Join actuals to a set of vehicle datapoints
+    def transform(self, vehicle_datapoints, y=None):
+        actuals = self._load_actuals()
+        actuals = actuals.rename(
+            {"trip_id": "gtfs_trip_id", "stop_id": "destination_gtfs_id"},
+            axis="columns"
+        )
+        actuals["gtfs_trip_id"] = actuals["gtfs_trip_id"].astype(str)
+        vehicle_datapoints["gtfs_trip_id"] = \
+            vehicle_datapoints["gtfs_trip_id"].astype(str)
+
+        merged_frame = vehicle_datapoints.merge(
+            actuals,
+            how="inner",
+            on=["gtfs_trip_id", "destination_gtfs_id"]
+        ).drop(
+            [
+                "generation",
+                "gtfs_trip_id",
+                "ocs_trip_id",
+                "vehicle_id_x",
+                "vehicle_id_y"
+            ],
+            axis=1
+        )
+
+        merged_frame["timestamp"] = merged_frame["timestamp"].apply(
+            lambda str: datetime.datetime.strptime(
+                str,
+                "%Y-%m-%dT%H:%M:%S.%fZ"
+            ).replace(tzinfo=pytz.UTC).timestamp()
+        )
+        merged_frame["actual_seconds_from_now"] = \
+            merged_frame["time"] - merged_frame["timestamp"]
+        merged_frame = merged_frame.drop(["timestamp", "time"], axis=1)
+
+        return merged_frame.dropna()
+
+    # Build a dataframe from prediction analyzer logs, dropping actuals without
+    # times, commuter rail trips, and duplicates (which we get because we log
+    # both from dev-green and prod).
+    def _load_actuals(self):
```

Comment:
Trying this out locally, I'm seeing a lot of negative actuals. I think, though I'm not positive, it's basically how long _ago_ did the vehicle leave the given stop ID to get to this location. For our purposes, though, it's not very useful. We could either try to reverse the data points to get actionable training examples (convert the GTFS stop IDs back to location IDs, convert the location IDs to GTFS stop IDs if applicable, and reverse them), or just drop all the negatives.

Reply:
Let's drop them for now, this stuff is complicated enough that I'm trying to err on the side of simplicity.